### ISSUE TITLE
regex.Matcher.start(group) and friends report wrong positions

### ISSUE THREAD

**Issue by MikaelMayer** [scala-js/scala-js#3134]
Using 0.6.19, the following code returns not the same result in scalajs vs. jvm

    "^(\\w+)(.*)".r.findAllMatchIn("x").toList.foreach { m =>
      for(i <- 1 to 2) {
        println("group " + i + ":" + m.start(i) + "-" + m.end(i))
      }
    }

On the JVM it prints as expected:

    group 1:0-1
    group 2:1-1

But on ScalaJS, it displays:

    group 1:0-1
    group 2:0-0

So the second group does not have the same offset. How can we prevent that?
Based on observation of the code here: https://github.com/scala-js/scala-js/blob/02d7f957ecc5eb86572bd72d14a2edc2aeba9498/javalib/src/main/scala/java/util/regex/Matcher.scala#L193 the ending position is the start plus the length of the group, so only the start position of the second group is actually wrong.

**Comment by sjrd** [comment]
The code is here: https://github.com/scala-js/scala-js/blob/0.6.x/javalib/src/main/scala/java/util/regex/Matcher.scala
It's a job for the 0.6.x branch.

Note that group-related features are *hacks*, because JS regexps don't provide that feature. In general you'll always be able to find a case where it breaks.

**Comment by MikaelMayer** [comment]
Ok, after playing around with the regular expressions, I found what actually groups are really doing.
m.start(i) in scalajs is the relative position of the start after the last end, whereas it is the absolute position in the JVM. If I can do a computation to fork my code between scalajs and JVM, I can work around that.

**Comment by sjrd** [comment]
If that's true, it's a bug that should be fixable in Scala.js. But I don't see how that can be the case from https://github.com/scala-js/scala-js/blob/72456873a56f1ea833be996bb57707f2faecf52a/javalib/src/main/scala/java/util/regex/Matcher.scala#L178-L188.

**Comment by MikaelMayer** [comment]
More precisely, after more investigations

* The position of the first and third group (and probably beyond) is absolute
* The position of the second group is _relative_ position iff the size of the second group is less or equal to the size of the first group.

I'll shortly post my experiments.

**Comment by MikaelMayer** [comment]
Actually, it's more complicated than I thought. I found the following difference tricky:
```scala
"^x(h+)  (h*) ".r.findAllMatchIn("xh  h ").toList.foreach { m =>
  for(i <- 1 to 2) {
    println("group " + i + ":" + m.start(i) + "-" + m.end(i))
  }
}

"^x(h+)  (g*) ".r.findAllMatchIn("xh  g ").toList.foreach { m =>
  for(i <- 1 to 2) {
    println("group " + i + ":" + m.start(i) + "-" + m.end(i))
  }
}
```

This prints:

    group 1:1-2
    group 2:1-2
    group 1:1-2
    group 2:4-5

Just by changing a letter from h to g changed the group number.

**Comment by MikaelMayer** [comment]
Another interesting case to compare. I add one more h in the second group of h. 

```scala
"^x(h+)   (h+) (h+) ".r.findAllMatchIn("xhhhhh   hhhhh h ").toList.foreach { m =>
  for(i <- 1 to 3) {
    println("group " + i + ":" + m.start(i) + "-" + m.end(i))
  }
}

"^x(h+)   (h+) (h+) ".r.findAllMatchIn("xhhhhh   hhhhhh h ").toList.foreach { m =>
  for(i <- 1 to 3) {
    println("group " + i + ":" + m.start(i) + "-" + m.end(i))
  }
}
```

The output is radically different for the second group only:

    group 1:1-6
    group 2:1-5
    group 3:1-2

    group 1:1-6
    group 2:9-15
    group 3:1-2

**Comment by MikaelMayer** [comment]
I observe the same behavior not just for repetitions, but this time when the second match is a *prefix* of the first.

    "^x(\\w+)   (\\w+) (h+) ".r.findAllMatchIn("xabcde   abcd h ").toList.foreach { m =>
      for(i <- 1 to 3) {
        println("group " + i + ":" + m.start(i) + "-" + m.end(i))
      }
    }

The second group is wrongly captured, but the third position is absolute

    group 1:1-6
    group 2:1-5
    group 3:14-15

. If I replace the last h by any of the strings a, ab, abc, abcd, abcde, I will obtain a relative position. However, for h replaced by abcdef, the absolute position is back. Another experiment explains why in the above the position starts at 1 for the second group.

    "^x(\\w+)   (\\w+) (\\w+) ".r.findAllMatchIn("xabcde   abgd abg ").toList.foreach { m =>
      for(i <- 1 to 3) {
        println("group " + i + ":" + m.start(i) + "-" + m.end(i))
      }
    }

The output is...

    group 1:1-6
    group 2:9-13
    group 3:9-12

So actually, it's the start position that is copied, from the previous string that the string matched is a prefix of. I don't know if this makes sense, but to me it clearer and clearer.
Replace "abg" by "abc" above and you'll get 1-4 for the third position.

**Comment by MikaelMayer** [comment]
Ok based on the line you provided, we can deduce how this arrived:

    groupStr => inputstr.indexOf(groupStr, last.index) 

If I'm not mistaken, it tries to look for the group string in the whole string starting at last.index, but last.index is always zero ! Why is that?

**Comment by MikaelMayer** [comment]
### Partial Solution

Ok, this is just because the hack tries to look for the position of the group in the original substring. Until we have complete solution, here is a better hack (modulo the fact that `last` can be handled like a regular array)

    ((-1) /: last.view.drop(1).take(group)) {
      case (lastIndex, groupStr) => inputstr.indexOf(groupStr, lastIndex)
    }

Here we take each matched group successively, and find their position. This will still fail for empty matches, but at this point, if we do not record more information, we cannot do better. At least this hack makes so that the matches are in order.

**Comment by MikaelMayer** [comment]
The partial solution is wrong for the following case:

    Pattern.compile("(Scala\\.js).*(Scala)$")
    .matcher("Scala.js is a Scalable javascript compiler based on Scala")

because it will think that the second match comes from the second occurrence of Scala in the string.

**Comment by sjrd** [comment]
Fixed in 4aecb56a7ca5c29d03a8ef7464e82f8d6a77f1b6.
